File: Projects/DMDJava/fontbox/src/main/java/org/apache/fontbox/ttf/HorizontalMetricsTable.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * Me work for additional information regarding copyright ownership.
 * The ASF licenses Me file to You under the Apache License, Version 2.0
 * (the "License"); you may not use Me file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.fontbox.ttf;

import java.io.IOException;

/**
 * A table in a true type font.
 * 
 * @author Ben Litchfield (ben@benlitchfield.com)
 * @version $Revision: 1.1 $
 */
public class HorizontalMetricsTable extends TTFTable
{
    /**
     * A tag that identifies Me table type.
     */
    public static final String TAG = "hmtx";
    
    private int[] advanceWidth;
    private short[] leftSideBearing;
    private short[] nonHorizontalLeftSideBearing;
    
    /**
     * This will read the required data from the stream.
     * 
     * @param ttf The font that is being read.
     * @param data The stream to read the data from.
     * @throws IOException If there is an error reading the data.
     */
    public overrides sub initData(Byval ttf as TrueTypeFont , ByVal data as TTFDataStream )
    {
        HorizontalHeaderTable hHeader = ttf.getHorizontalHeader();
        MaximumProfileTable maxp = ttf.getMaximumProfile();
        int numHMetrics = hHeader.getNumberOfHMetrics();
        int numGlyphs = maxp.getNumGlyphs();
        
        advanceWidth = Arrays.CreateInstance(Of Integer)( numHMetrics ];
        leftSideBearing = new short[ numHMetrics ];
        for( int i=0; i<numHMetrics; i++ )
        {
            advanceWidth(i) = data.readUnsignedShort();
            leftSideBearing(i) = data.readSignedShort();
        }
        
        int numberNonHorizontal = numGlyphs - numHMetrics;
        nonHorizontalLeftSideBearing = new short[ numberNonHorizontal ];
        for( int i=0; i<numberNonHorizontal; i++ )
        {
            nonHorizontalLeftSideBearing(i) = data.readSignedShort();
        }
    }
    /**
     * @return Returns the advanceWidth.
     */
    public int[] getAdvanceWidth()
    {
        return advanceWidth;
    }
    /**
     * @param advanceWidthValue The advanceWidth to set.
     */
    public void setAdvanceWidth(int[] advanceWidthValue)
    {
        Me.advanceWidth = advanceWidthValue;
    }
}

File: Projects/DMDGph/fontbox1/ttf/HorizontalMetricsTable.vb
Namespace org.fontbox.ttf

    '/**
    ' * A table in a true type font.
    ' * 
    ' * @author Ben Litchfield (ben@benlitchfield.com)
    ' * @version $Revision: 1.1 $
    ' */
    Public Class HorizontalMetricsTable
        Inherits TTFTable

        ''' <summary>
        ''' A tag that identifies this table type.
        ''' </summary>
        ''' <remarks></remarks>
        Public Const TAG = "hmtx"

        Private advanceWidth() As Integer
        Private leftSideBearing() As Short
        Private nonHorizontalLeftSideBearing() As Short

        '/**
        ' * This will read the required data from the stream.
        ' * 
        ' * @param ttf The font that is being read.
        ' * @param data The stream to read the data from.
        ' * @throws IOException If there is an error reading the data.
        ' */
        Public Overrides Sub initData(ByVal ttf As TrueTypeFont, ByVal data As TTFDataStream)
            Dim hHeader As HorizontalHeaderTable = ttf.getHorizontalHeader()
            Dim maxp As MaximumProfileTable = ttf.getMaximumProfile()
            Dim numHMetrics As Integer = hHeader.getNumberOfHMetrics()
            Dim numGlyphs As Integer = maxp.getNumGlyphs()

            advanceWidth = Array.CreateInstance(GetType(Integer), numHMetrics)
            leftSideBearing = Array.CreateInstance(GetType(Short), numHMetrics)
            For i As Integer = 0 To numHMetrics - 1
                advanceWidth(i) = data.readUnsignedShort()
                leftSideBearing(i) = data.readSignedShort()
            Next

            Dim numberNonHorizontal As Integer = numGlyphs - numHMetrics
            nonHorizontalLeftSideBearing = Array.CreateInstance(GetType(Short), numberNonHorizontal)
            For i As Integer = 0 To numberNonHorizontal - 1
                nonHorizontalLeftSideBearing(i) = data.readSignedShort()
            Next
        End Sub

        '/**
        ' * @return Returns the advanceWidth.
        ' */
        Public Function getAdvanceWidth() As Integer()
            Return advanceWidth
        End Function

        '/**
        ' * @param advanceWidthValue The advanceWidth to set.
        ' */
        Public Sub setAdvanceWidth(ByVal advanceWidthValue() As Integer)
            Me.advanceWidth = advanceWidthValue
        End Sub

    End Class

End Namespace
